rules/prefer-class-fields.js

```diff
@@ -0,0 +1,182 @@
+	const isOnlyNodeOnLine = lineText.trim() === nodeText;
+
+	return isOnlyNodeOnLine
+		? fixer.removeRange([
+			sourceCode.getIndexFromLoc({line, column: 0}),
+			sourceCode.getIndexFromLoc({line: line + 1, column: 0}),
+		])
+		: fixer.remove(node);
+};
+
+/**
+@param {string} propertyName
+@param {import('estree').ClassBody} classBody
+*/
+const findClassFieldNamed = (propertyName, classBody) => {
+	for (const classBodyChild of classBody.body) {
+		if (
+			classBodyChild.type === 'PropertyDefinition'
+			&& classBodyChild.key.type === 'Identifier'
+			&& classBodyChild.key.name === propertyName
+		) {
+			return classBodyChild;
+		}
+	}
+};
+
+/**
+@param {string} propertyName
+@param {string} propertyValue
+@param {import('estree').ClassBody} classBody
+@param {import('estree').MethodDefinition} constructor
+@param {import('eslint').Rule.RuleContext['sourceCode']} sourceCode
+@param {import('eslint').Rule.RuleFixer} fixer
+*/
+const addOrReplaceClassFieldDeclaration = (
+	propertyName,
+	propertyValue,
+	classBody,
+	constructor,
+	sourceCode,
+	fixer,
+) => {
+	const alreadyExistingDeclaration = findClassFieldNamed(
+		propertyName,
+		classBody,
+	);
+
+	if (alreadyExistingDeclaration) {
+		return fixer.replaceText(
+			alreadyExistingDeclaration,
+			`${propertyName} = ${propertyValue};`,
+		);
+	}
+
+	const classBodyStartRange = [classBody.range[0], classBody.range[0] + 1];
+	const indent = getIndentString(constructor, sourceCode);
+	return fixer.insertTextAfterRange(
+		classBodyStartRange,
+		`\n${indent}${propertyName} = ${propertyValue};`,
+	);
+};
+
+/**
+@type {import('eslint').Rule.RuleModule['create']}
+*/
+const create = context => {
+	const {sourceCode} = context;
+
+	return {
+		ClassBody(classBody) {
+			const constructor = classBody.body.find(x => x.kind === 'constructor');
+
+			if (!constructor || constructor.type !== 'MethodDefinition') {
+				return;
+			}
+
+			const constructorBody = constructor.value.body?.body;
+
+			if (!constructorBody) {
+				return;
+			}
+
+			const firstInvalidProperty = constructorBody.findIndex(
+				node => node.type !== 'ExpressionStatement',
+			);
+			const validConstructorProperties
+				= firstInvalidProperty === -1
+					? constructorBody
+					: constructorBody.slice(0, firstInvalidProperty);
+
+			for (
+				let index = validConstructorProperties.length - 1;
+				index >= 0;
+				index--
+			) {
+				const node = validConstructorProperties[index];
+				if (
+					isThisAssignmentExpression(node)
+					&& node.expression.right?.type === 'Literal'
+					&& node.expression.operator === '='
+					&& node.expression.left.property.type === 'Identifier'
+					&& !node.expression.left.computed
+				) {
+					return {
+						node,
+						messageId: MESSAGE_ID,
+
+						/**
+						@param {import('eslint').Rule.RuleFixer} fixer
+						*/
+						* fix(fixer) {
+							yield removeFieldAssignment(node, sourceCode, fixer);
+							yield addOrReplaceClassFieldDeclaration(
```

Comment:
Let's not remove code, especially there can be side effects. We can use suggestions or simply just report error, if property already exists.

Reply:
My thinking was: property in the constructor always takes priority over class field declaration, so there are no side-effects. IMO it's a safe assumption.
WDYT?

Reply:
```js
const getFoo = () => console.log('called')

new class {
  foo = getFoo();  <-- this function call is removed.
  constructor() {
     this.foo = 1
  }
}
```

Reply:
You're right, I've rewritten it as a suggestion 3153a8922da7aa79d28680302dfc186b560bd6e0